### ISSUE TITLE
Import Peru case data

### ISSUE THREAD

**Issue by z023** [globaldothealth/list#1016]
Uploaded Peru automated source, triggered retrieval and received this error:
Error: Request failed with status code 500

**Comment by attwad** [comment]
Seems like the data is large and the ingestion function needs more memory to fetch it.

```
[ERROR] MemoryError
Traceback (most recent call last):
  File "/var/task/retrieval.py", line 237, in lambda_handler
    file_name, s3_object_key = retrieve_content(
  File "/var/task/retrieval.py", line 89, in retrieve_content
    r = requests.get(url, headers=headers)
  File "/var/task/requests/api.py", line 76, in get
    return request('get', url, params=params, **kwargs)
  File "/var/task/requests/api.py", line 61, in request
    return session.request(method=method, url=url, **kwargs)
  File "/var/task/requests/sessions.py", line 530, in request
    resp = self.send(prep, **send_kwargs)
  File "/var/task/requests/sessions.py", line 683, in send
    r.content
  File "/var/task/requests/models.py", line 829, in content
    self._content = b''.join(self.iter_content(CONTENT_CHUNK_SIZE)) or b''
```

**Comment by Mougk** [comment]
Just tried again using trigger retrieval. Still does not work (data file has about 729k cases).

**Comment by attwad** [comment]
The OOM is gone, the parser runs this time but it seems it needs to be made more robust:

```
[ERROR] ValueError: could not convert string to float: ''
Traceback (most recent call last):
  File "/var/task/peru.py", line 95, in lambda_handler
    return parsing_lib.run_lambda(event, context, parse_cases)
  File "/opt/python/parsing_lib.py", line 250, in run_lambda
    common_lib.complete_with_error(
  File "/opt/python/common_lib.py", line 88, in complete_with_error
    raise exception
  File "/opt/python/parsing_lib.py", line 237, in run_lambda
    count_created, count_updated = write_to_server(
  File "/opt/python/parsing_lib.py", line 110, in write_to_server
    batch = batch_of(cases, CASES_BATCH_SIZE)
  File "/opt/python/parsing_lib.py", line 96, in batch_of
    batch.append(next(cases))
  File "/opt/python/parsing_lib.py", line 78, in prepare_cases
    for case in cases:
  File "/var/task/peru.py", line 85, in parse_cases
    "start": float(entry["EDAD"]),
```

**Comment by attwad** [comment]
Felix can you look at this please?

**Comment by z023** [comment]
Only 147k ingested when there should be 800k. Is this related to issue #1124?

**Comment by attwad** [comment]
@z023 Yes Peru cannot be imported at once, it needs to be done by the backfill feature @axmb added, he's trying it out in dev today and will get back to you for how to run it in prod once we confirmed it works as intended.

**Comment by z023** [comment]
Ah ok thanks. I just noticed the new automated backfill option.

**Comment by axmb** [comment]
Backfill's ready for use -- there's [a bit of user documentation on it here](https://docs.google.com/document/d/1tK1PpAqQ37bBVap0g-evGlc06hcnvBpfE6lvpGyu8rc/edit#heading=h.imiskg63rzz9). Feel free to give it a shot and poke me with any questions.

That said, tomorrow Allyson/Stephen/I will be working on merging legacy/new data, so any manual backfilling may need to be redone.

**Comment by Mougk** [comment]
@sratcliffe118 @z023 is this one ready, can we close?

**Comment by sratcliffe118** [comment]
Yes this is fully backfilled, we can close

**Comment by z023** [comment]
I was checking through the ingested cases for verification and picked a few unique IDs at random and noticed the demographic information being ingested is totally different to what it should be from looking at the original dataset. Also confirmed by Moritz and David.

Here is an example of one ID and what has been ingested.
<img width="1005" alt="Screenshot 2020-10-20 at 17 17 05" src="https://user-images.githubusercontent.com/65172419/96615364-15648d00-12f9-11eb-974a-ec1b6776a12e.png">

I then looked up the original source entry ID (not case ID) in the original dataset. Here is what the original dataset has for this ID
![Screenshot 2020-10-20 at 17 18 32](https://user-images.githubusercontent.com/65172419/96615401-20b7b880-12f9-11eb-9f4e-aebbb9a60a5b.png)

For this one the age and confirmation date don't match. On others I've also seen the geo locations and gender don't match.

**Comment by calremmel** [comment]
Peru is now also throwing an error, which I'll include below. I'll give this all a look today, Please let me know if you have any thoughts!

```
[ERROR] KeyError: 'UUID'
Traceback (most recent call last):
  File "/var/task/peru.py", line 100, in lambda_handler
    return parsing_lib.run_lambda(event, context, parse_cases)
  File "/opt/python/parsing_lib.py", line 304, in run_lambda
    common_lib.complete_with_error(
  File "/opt/python/common_lib.py", line 98, in complete_with_error
    raise exception
  File "/opt/python/parsing_lib.py", line 289, in run_lambda
    count_created, count_updated = write_to_server(
  File "/opt/python/parsing_lib.py", line 114, in write_to_server
    batch = batch_of(cases, cases_batch_size)
  File "/opt/python/parsing_lib.py", line 98, in batch_of
    batch.append(next(cases))
  File "/opt/python/parsing_lib.py", line 227, in <genexpr>
    return (case for case in case_data if case_is_within_range(case, cutoff_date, op))
  File "/opt/python/parsing_lib.py", line 80, in prepare_cases
    for case in cases:
  File "/var/task/peru.py", line 79, in parse_cases
    "sourceEntryId": entry["UUID"],
```

**Comment by z023** [comment]
@fojackson8 Could you take a look at the parser code? Peru is ingesting fine but the demographic information is coming through totally different to the original dataset when comparing UUIDs, as indicated above.

**Comment by fojackson8** [comment]
@z023 good spot. The dataset was changed recently on their side which caused the parser to break. I updated the parser, but have compared the most recent dataset to the previous one and indeed the UUIDs correspond to different cases in the two datasets. So I don't think its a parser error, but just reflects a UUID update from the source. It does however call the UUIDs into question..
@Mougk @calremmel for visibility
